src/resource_limiter.rs

```diff
@@ -0,0 +1,174 @@
+        self.measure(false);
+        if self.current_usage.exceeds(&self.limits) {
+            error!( target: "SCP",
+                "Resource limits exceeded -- Time elapsed: {} ms, Time limit: {} ms; Memory usage: {} bytes, Memory limit: {} bytes",
+                self.current_usage.time.as_millis(), self.limits.time.as_millis(), self.current_usage.mem_bytes, self.limits.mem_bytes
+            );
+            return Err(FbasError::ResourcelimitExceeded(self.current_usage));
+        }
+        Ok(())
+    }
+}
+
+impl ResourceLimiter {
+    pub fn new(time_limit_ms: u64, memory_limit_bytes: usize) -> Self {
+        Self(Rc::new(RefCell::new(ResourceLimiterImpl::new(
+            time_limit_ms,
+            memory_limit_bytes,
+        ))))
+    }
+
+    pub fn unlimited() -> Self {
+        Self(Rc::new(RefCell::new(ResourceLimiterImpl::new(
+            u64::MAX,
+            usize::MAX,
+        ))))
+    }
+
+    pub fn measure(&self, verbose: bool) {
+        self.0.borrow_mut().measure(verbose);
+    }
+
+    pub fn measure_and_enforce_limits(&self) -> Result<(), FbasError> {
+        self.0.borrow_mut().measure_and_enforce_limits()
+    }
+
+    pub fn get_time_ms(&self) -> u64 {
+        self.0.borrow().current_usage.time.as_millis() as u64
+    }
+
+    pub fn get_mem_bytes(&self) -> usize {
+        self.0.borrow().current_usage.mem_bytes
+    }
+}
+
+impl Callbacks for ResourceLimiter {
+    fn on_start(&mut self) {
+        self.measure(true);
+        trace!( target: "SCP",
+            "c ============================[ Search Statistics ]=============================="
+        );
+        trace!( target: "SCP",
+            "c | Conflicts |          ORIGINAL         |          LEARNT          | Progress |"
+        );
+        trace!( target: "SCP",
+            "c |           |    Vars  Clauses Literals |    Limit  Clauses Lit/Cl |          |"
+        );
+        trace!( target: "SCP",
+            "c ==============================================================================="
+        );
+    }
+
+    fn on_result(&mut self, _: lbool) {
+        trace!( target: "SCP",
+            "c ==============================================================================="
+        );
+        self.measure(true);
+    }
+
+    fn on_progress<F>(&mut self, p: F)
+    where
+        F: FnOnce() -> ProgressStatus,
+    {
+        let p = p();
+        trace!( target: "SCP",
+            "c | {:9} | {:7} {:8} {:8} | {:8} {:8} {:6.0} | {:6.3} % |",
+            p.conflicts,
+            p.dec_vars,
+            p.n_clauses,
+            p.n_clause_lits,
+            p.max_learnt,
+            p.n_learnt,
+            p.n_learnt_lits,
+            p.progress_estimate
+        );
+    }
+
+    fn on_gc(&mut self, old: usize, new: usize) {
+        trace!( target: "SCP",
+            "|  Garbage collection:   {:12} bytes => {:12} bytes             |",
+            old, new
+        );
+    }
+
+    fn stop(&self) -> bool {
```

Comment:
I'm curious why this worked better than the previous approach -- I think you said it does? It's still _basically_ doing the same thing though, right? Counting allocations at callbacks and telling the solver to stop when a limit is exceeded? What works better this time? (I don't mind this code, it looks fine, I'm just not clear what changed/improved)